Review 2

TheLosst opened the pull request:
Исправление импортов **useSelector** и **useDispatch**
Исправление генерации **UUID в редьюсере**
Настройка **модальной** **логики** роутинга
Исправление логики перезагрузки страницы

Исправлены **проблемы** **с** получением данных из **.env**

Commit:
Шаг 1. Роутинг

Commit:
Запросы к серверу НАЧАЛО КОНЦА

Commit:
Продолжаем клоунфиесту с реактом)

Commit:
Двигаемся вперед

Commit:
Готов в ревью

Commit:
Исправление импортов useSelector и useDispatch

Исправление генерации UUID в редьюсере

Настройка модальной логики роутинга

Исправление логики перезагрузки страницы

Commit:
Исправлено НАЗВАНИЕ переменной в burger-api (из-за этого собственно и не работало, ибо я назвал ее криво...)

gennady-bars commented on src/pages/login/login.tsx:
```diff
@@ -1,17 +1,41 @@
 import { FC, SyntheticEvent, useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
```
* Хуки `useSelector` и `useDispatch` нужно импортировать из файла `store.ts` (они там типизированные уже), а не из библиотеки `react-redux`.  Нужно найти через поиск `react-redux` и заменить **все импорты**.

gennady-bars commented on src/components/app-header/app-header.tsx:
```diff
@@ -1,4 +1,76 @@
+export const AppHeader: FC = () => {
+  const user = useSelector((state) => state.user);
+  const location = useLocation();
+
```
**Можно лучше**


Если интересно, то можно создать **универсальную** функцию запроса `request` , которая будет содержать в себе уже **базовый урл**, проверку на **ok** и проверку на `success`, чтобы не дублировать их в каждом запросе.

```js

// 1 раз объявляем базовый урл
export const BASE_URL = "https://norma.nomoreparties.space/api/";

// создаем функцию проверки ответа на `ok`
const checkResponse = (res) => {
  if (res.ok) {
    return res.json();
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ошибка ${res.status}`);
};

// создаем функцию проверки на `success`
const checkSuccess = (res) => {
  if (res && res.success) {
    return res;
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ответ не success: ${res}`);
};

// создаем универсальную фукнцию запроса с проверкой ответа и `success`
// В вызов приходит `endpoint`(часть урла, которая идет после базового) и опции
const request = (endpoint, options) => {
  // а также в ней базовый урл сразу прописывается, чтобы не дублировать в каждом запросе
  return fetch(`${BASE_URL}${endpoint}`, options)
    .then(checkResponse)
    .then(checkSuccess);
};

```


И теперь не нужно будет дублировать все это в каждом запросе. Нужно просто заменить в запросах все `fetch` на `request`, удалив лишние проверки и базовый урл. Все остальное останется без изменений

Код станет намного чище. Будет мелкий аналог `axios`

Пример запроса ингредиентов

```js
const getIngredients =  () => request("ingredients");

```

В `get`-запросах даже не нужно указывать 2й параметр. Он получится `undefined`, что без проблем для `fetch` (который внутри).

gennady-bars commented on src/pages/login/login.tsx:
```diff
@@ -1,17 +1,41 @@
+import { TRootReducer } from '../../services/types';
 
 export const Login: FC = () => {
   const [email, setEmail] = useState('');
   const [password, setPassword] = useState('');
 
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,171 @@
+
+  useEffect(() => {
+    // Загружаем ингредиенты при старте приложения
+    dispatch(fetchIngredients());
```
`dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.tsx:
```diff
@@ -27,6 +36,13 @@ export const BurgerIngredients: FC = () => {
+  // Загружаем ингредиенты при монтировании компонента
+  useEffect(() => {
+    if (ingredientsStatus === 'idle') {
+      dispatch(fetchIngredients());
```
`dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/redux-demo/redux-demo.tsx:
```diff
@@ -0,0 +1,264 @@
+  useEffect(() => {
+    // Загружаем ингредиенты
+    if (ingredientsStatus === 'idle') {
+      dispatch(fetchIngredients());
```
`dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/pages/constructor-page/constructor-page.tsx:
```diff
@@ -1,19 +1,32 @@
+  );
+  const dispatch = useDispatch();
+  useEffect(() => {
+    dispatch(fetchIngredients());
```
`dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,171 @@
+  return (
+    <div className={styles.app}>
+      <AppHeader />
+      <Routes>
```
Вам нужно использовать средства **роутинга** для определения, показывать ли попап с деталями, или показывать страницу. Вам должны были дать ссылку на документацию по этой логике. Вот она https://v5.reactrouter.com/web/example/modal-gallery . Пожалуйста, посоветуйтесь с наставником. В 6й версии роутинга примерно такая же логика.


Вся логика уже была с `background` в стартер-ките

` background` в 6й версии роутинга можно определять и передавать в ` Routes` вот так:
 
```jsx
  const background = location.state?.background

  return (
      <Routes location={background || location}>

```

gennady-bars replied:
Это решит вопрос с отображением страницы под попапом

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,171 @@
+          </ProtectedRoute>
+        )}
+      {location.pathname.includes('/ingredients/') &&
+        location.pathname.match(/\/ingredients\/[^/]+$/) && <IngredientModal />}
```
Все модальные окна будут во 2м `Routes` при наличии `background`.

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)


Работа проделана огромная, но есть некоторые недочеты: 




* Хуки `useSelector` и `useDispatch` нужно импортировать из файла `store.ts` (они там типизированные уже), а не из библиотеки `react-redux`.  Нужно найти через поиск `react-redux` и заменить **все импорты**.
* Не открывается попап ингредиента 
* По заданию нужно сделать логику, при которой бы показывался попап с деталями ингредиента, если кликнуть по ингредиенту, но если скопировать ссылку на ингредиент при открытом попапе и **вставить в окно браузера**, то должна уже показываться **страница** с этими деталями. Такая же логика должна быть и для заказов ленты и истории заказов
* Если **перезагрузить** страницу с открытым попапом заказа **истории**, то этот открытый попап опять должен появляться (или страница с деталями). А сейчас ничего не отображается после перезагрузки. `ProtectedRoute` нужно использовать для попапа **истории заказов**, так как его могут посмотреть только **авторизованные** пользователи
* Нет **автологина**. Если перезагрузить страницу в профиле, то я опять оказываюсь на странице логина, а должен быть автологин и я должен попадать туда, откуда перезагружал страницу
* `dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.
* Все модальные окна будут во 2м `Routes` при наличии `background`.


**Можно лучше**

* Желательно добавить **прелоадер** загрузки данных пользователя в `ProtectedRoute`, чтобы во  время ожидания ответа от севрера у авторизированного пользователя не показывалась страница **логина**.

Commit:
Исправления по пунктам:
1. src/pages/login/login.tsx
Хуки useSelector и useDispatch нужно импортировать из файла store.ts (они там типизированные уже), а не из библиотеки react-redux. Нужно найти через поиск react-redux и заменить все импорты.

2. src/components/app/app.tsx
dispatch(fetchIngredients()) достаточно 1 раз вызвать в App, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

3. src/components/burger-ingredients/burger-ingredients.tsx
dispatch(fetchIngredients()) достаточно 1 раз вызвать в App, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

4. src/components/redux-demo/redux-demo.tsx
dispatch(fetchIngredients()) достаточно 1 раз вызвать в App, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

5. src/pages/constructor-page/constructor-page.tsx
dispatch(fetchIngredients()) достаточно 1 раз вызвать в App, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

6. src/components/app/app.tsx
Вам нужно использовать средства роутинга для определения, показывать ли попап с деталями, или показывать страницу. Вам должны были дать ссылку на документацию по этой логике. Вот она https://v5.reactrouter.com/web/example/modal-gallery . Пожалуйста, посоветуйтесь с наставником. В 6й версии роутинга примерно такая же логика.

Вся логика уже была с background в стартер-ките

 background в 6й версии роутинга можно определять и передавать в  Routes вот так:
 ```
   const background = location.state?.background

  return (
      <Routes location={background || location}>
```
Это решит вопрос с отображением страницы под попапом

7. src/components/app/app.tsx
Все модальные окна будут во 2м Routes при наличии background.
